File: AlgorithmicToolbox/week5_dynamic_programming1/3_edit_distance/3_edit_distance.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using std::string;
using std::vector;
/*
The edit distance between two strings is the minimum number of operations (insertions, deletions,
and substitutions of symbols) to transform one string into another. It is a measure of similarity
of two strings. Edit distance has applications, for example, in computational biology, natural
language processing, and spell checking. Your goal in this problem is to compute the edit distance
between two strings.
*/
int edit_distance(const string& str1, const string& str2) {
	vector<int> d(str2.size()+1);
	vector<vector<int> > D(str1.size()+1, d);
	for (auto i = 0; i != str1.size() + 1; ++i) {
		D[i][0] = i;
	}
	for (auto j = 0; j != str2.size() + 1; ++j) {
		D[0][j] = j;
	}
	for (auto i = 1; i != str1.size() + 1; ++i) {
		for (auto j = 1; j != str2.size() + 1; ++j) {
			int insertion = D[i][j - 1] + 1;
			int deletion = D[i - 1][j] + 1;
			int match = D[i - 1][j - 1];
			int mismatch = D[i - 1][j - 1] + 1;

			if (str1[i - 1] == str2[j - 1]) {
				D[i][j] = std::min({ insertion, deletion, match });
			}
			else {
				D[i][j] = std::min({ insertion, deletion, mismatch });
			}
		}
	}
	return D[str1.size()][str2.size()];
}

int main() {
	string str1;
	string str2;
	std::cin >> str1 >> str2;
	std::cout << edit_distance(str1, str2) << std::endl;
	return 0;
}

File: AlgorithmicToolbox/week5_dynamic_programming1/3_edit_distance/3_edit_distance.py
#!/usr/bin/env python
# coding: utf-8

# In[ ]:


# python3
import numpy as np

"""
The edit distance between two strings is the minimum number of operations (insertions, deletions,
and substitutions of symbols) to transform one string into another. It is a measure of similarity
of two strings. Edit distance has applications, for example, in computational biology, natural 
language processing, and spell checking. Your goal in this problem is to compute the edit distance
between two strings.
"""

def edit_distance(first_string, second_string):
    
    A, B = first_string, second_string
    
    D = np.zeros((len(A)+1, len(B)+1))
    
    for i in range(len(A)+1):
        D[i][0] = i
    for j in range(len(B)+1):
        D[0][j] = j
    
    for j in range(1, len(B)+1):
        for i in range(1, len(A)+1):
            insertion = D[i][j-1] + 1
            deletion = D[i-1][j] + 1
            match = D[i-1][j-1]
            mismatch = D[i-1][j-1] + 1
            if A[i-1] == B[j-1]:
                D[i][j] = min(insertion, deletion, match)
            else:
                D[i][j] = min(insertion, deletion, mismatch)
    
    
    return int(D[len(A)][len(B)])


if __name__ == "__main__":
    print(edit_distance(input(), input()))


# In[ ]:




